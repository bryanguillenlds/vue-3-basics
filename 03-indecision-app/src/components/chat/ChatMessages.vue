<template>
	<div class="flex-1 overflow-y-auto p-4">
		<div class="flex flex-col space-y-2">
			<!-- Messages go here -->
			<!-- Example Message -->
			<ChatBubble
				v-for="message in messages"
				:key="message.id"
				:message="message.message"
				:sentByMe="message.sentByMe"
				:image="message.image"
			/>
		</div>
	</div>
</template>

<script setup lang="ts">
import type { ChatMessage } from '@/interfaces/chat-message-interface';
import ChatBubble from '@/components/chat/chatBubble.vue';

interface Props {
	messages: ChatMessage[];
}

const props = defineProps<Props>();
</script>
