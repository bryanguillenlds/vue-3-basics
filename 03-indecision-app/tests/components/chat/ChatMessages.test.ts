import { mount } from '@vue/test-utils';
import ChatMessages from '@/components/chat/ChatMessages.vue';
import { vi } from 'vitest';
import { nextTick } from 'vue';

const messages = [
	{
		id: 1,
		message: 'Hello, world!',
		sentByMe: true,
		image: null
	},
	{
		id: 2,
		message: 'Hello, world!',
		sentByMe: false,
		image: 'https://via.placeholder.com/150'
	}
];

describe('ChatMessages', () => {
	const wrapper = mount(ChatMessages, {
		props: {
			messages
		}
	});

	test('should render the chat messages component correctly', () => {
		const chatBubbles = wrapper.findAllComponents({ name: 'ChatBubble' });

		expect(wrapper.html()).toMatchSnapshot();

		expect(chatBubbles.length).toBe(messages.length);
	});

	test('should scroll to the bottom when a new message is added', async () => {
		const scrollToMock = vi.fn();

		// Get the chat container reference from the component
		// This is the div that will be scrolled
		const chatRef = wrapper.vm.$refs.chatRef as HTMLDivElement;

		// Replace the real scrollTo method with our mock function
		// This allows us to verify that scrollTo was called without actually scrolling
		chatRef.scrollTo = scrollToMock;

		// Simulate adding a new message by changing the props
		// IMPORTANT: This creates a NEW array at a different memory location
		// That's why we need the getter function in the watch() - direct watching would miss this change (this is in the component)
		wrapper.setProps({
			messages: [
				...messages,
				{
					id: 3,
					message: 'Hey there',
					sentByMe: true,
					image: null
				}
			]
		});

		// Wait for the setTimeout in the component to complete
		// The component uses a 100ms delay to ensure DOM is fully updated before scrolling
		await new Promise((resolve) => setTimeout(resolve, 150));

		//Expect the scroll top to be the same as the scroll height
		expect(chatRef.scrollTop).toBe(chatRef.scrollHeight);
		// Verify that scrollTo was called exactly once
		expect(scrollToMock).toHaveBeenCalledTimes(1);

		// Verify that scrollTo was called with the correct parameters
		// The component should scroll to the bottom (scrollHeight) with smooth behavior
		expect(scrollToMock).toHaveBeenCalledWith({
			top: chatRef.scrollHeight,
			behavior: 'smooth'
		});

		//Expect the scroll top to be the same as the scroll height
		expect(chatRef.scrollTop).toBe(chatRef.scrollHeight);
	});
});
