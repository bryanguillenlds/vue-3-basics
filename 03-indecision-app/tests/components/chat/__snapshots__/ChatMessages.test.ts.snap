// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatMessages > should render the chat messages component correctly 1`] = `
"<div class="flex-1 overflow-y-auto p-4">
  <div class="flex flex-col space-y-2">
    <!-- Messages go here -->
    <!-- Example Message -->
    <div class="flex justify-end">
      <div class="bg-blue-200 text-black p-2 rounded-lg max-w-xs">Hello, world!</div>
    </div><!-- Example Received Message -->
    <div class="flex">
      <div class="bg-gray-300 text-black p-2 rounded-lg max-w-xs"><span class="capitalize">Hello, world!</span><img src="https://via.placeholder.com/150" alt="YesNoImage" class="w-52 h-52 object-cover rounded-md"></div>
    </div>
  </div>
</div>"
`;
